snorkel/learning/logistic_regression.py

```diff
@@ -109,3 +117,90 @@ def save_info(self, model_name):
     def load_info(self, model_name):
         with open('{0}.info'.format(model_name), 'rb') as f:
             self.d, self.lr, self.l1_penalty, self.l2_penalty = cPickle.load(f)
+
+
+class SparseLogisticRegression(LogisticRegression):
+
+    def __init__(self, save_file=None, name='SparseLR'):
+        """Sparse noise-aware logistic regression in TensorFlow"""
+        self.indices = None
+        self.shape   = None
+        self.ids     = None
+        self.weights = None
+        super(SparseLogisticRegression, self).__init__(
+            save_file=save_file, name=name
+        )
+
+    def _build(self):
+        # Define input placeholders
+        self.indices = tf.placeholder(tf.int64) 
+        self.shape   = tf.placeholder(tf.int64, (2,))
+        self.ids     = tf.placeholder(tf.int64)
+        self.weights = tf.placeholder(tf.float32)
+        self.Y       = tf.placeholder(tf.float32, (None, 1))
+        # Define training variables
+        sparse_ids = tf.SparseTensor(self.indices, self.ids, self.shape)
+        sparse_vals = tf.SparseTensor(self.indices, self.weights, self.shape)
+        w = tf.Variable(tf.random_normal((self.d, 1), mean=0, stddev=0.01))
+        b = tf.Variable(tf.random_normal((1, 1), mean=0, stddev=0.01))
+        z = tf.nn.embedding_lookup_sparse(
+            params=w, sp_ids=sparse_ids, sp_weights=sparse_vals, combiner='sum'
+        )
+        h = tf.add(z, b)
+        # Build model
+        self.loss = tf.reduce_sum(
+            tf.nn.sigmoid_cross_entropy_with_logits(h, self.Y)
+        )
+        self.train_fn = tf.train.ProximalGradientDescentOptimizer(
+            learning_rate=tf.cast(self.lr, dtype=tf.float32),
+            l1_regularization_strength=tf.cast(self.l1_penalty, tf.float32),
+            l2_regularization_strength=tf.cast(self.l2_penalty, tf.float32),
+        ).minimize(self.loss)
+        self.prediction = tf.nn.sigmoid(h)
+        self.save_dict = {'w': w, 'b': b}
+        # Get nnz operation
+        self.nnz = tf.reduce_sum(tf.cast(
+            tf.not_equal(w, tf.constant(0, tf.float32)), tf.int32
+        ))
+
+    def _batch_sparse_data(self, X):
+        """Convert sparse batch matrix to sparse inputs for embedding lookup"""
+        if not issparse(X):
+            raise Exception("Matrix X must be scipy.sparse type")
+        X_lil = X.tolil()
+        indices, ids, weights = [], [], []
+        max_len = 0
+        for i, (row, data) in enumerate(zip(X_lil.rows, X_lil.data)):
+            max_len = max(max_len, len(row))
+            indices.extend((i, t) for t in xrange(len(row)))
+            ids.extend(row)
+            weights.extend(data)
+        shape = (len(X_lil.rows), max_len)
+        return indices, shape, ids, weights
+
+    def _run_batch(self, X_train, y_train, i, r):
+        """Run a single batch update"""
+        # Get batch sparse tensor data
+        indices, shape, ids, weights = self._batch_sparse_data(X_train[i:r, :])
```

Comment:
Is it worth converting `X_train` to CSR first before doing these cuts for each batch?